### PR TITLE
Increase the duration to expire an assumed pod

### LINKED ISSUE #106361: Scheduler will run into race conditions on large scale clusters
### What happened?

The scheduler has a [30s timeout](https://github.com/kubernetes/kubernetes/blob/0fefe4e605ba05793e919598c09390769c127b3c/pkg/scheduler/scheduler.go#L231) for the bind operation to succeed; if we don't get a response within 30s, the in-memory assignment of pod to node in the [scheduler cache expires](https://github.com/kubernetes/kubernetes/blob/0fefe4e605ba05793e919598c09390769c127b3c/pkg/scheduler/internal/cache/cache.go#L720).

A race condition will happen in the follow case:

1. pod1 is assigned to a node, scheduler cache is updated with the assignment, bind operation issued to apiserver.

1. if the apiserver is under huge pressure, bind takes more than 30s, scheduler expires the cached pod-to-node assignment.

1. bind eventually succeeds, but because the apiserver is under huge pressure, the pod update with the node name takes a long time to propagate to the scheduler.

1. because the pod update took a long time to propagate and the cache entry expired, the scheduler is not aware that the assignment actually happened, and so it had no problem assigning a second pod to the same node that would otherwise not fit if the scheduler was aware that the first pod was eventually assigned to the node.

On the scheduler side, what we need to do is make the 30s longer for large clusters, and ideally adaptable to cluster state.

/sig scheduling

### What did you expect to happen?

No race conditions.

### How can we reproduce it (as minimally and precisely as possible)?

Create a large scale cluster.

### Anything else we need to know?

_No response_

### Kubernetes version

<details>

```console
$ kubectl version
# paste output here
```

</details>

### Cloud provider

<details>

</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

### PR DESCRIPTION
**Pull request by ahg-g**
#### What type of PR is this?

/kind bug

#### Which issue(s) this PR fixes:
Part of #106361

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

### CODE REVIEW

File: pkg/scheduler/scheduler.go [hunk 1]
```diff
@@ -57,6 +57,8 @@ const (
 	SchedulerError = "SchedulerError"
 	// Percentage of plugin metrics to be sampled.
 	pluginMetricsSamplePercent = 10
+	// Duration the scheduler will wait before expiring an assumed pod.
+	durationToExpireAssumedPod = 15 * time.Minute
```
Comment (Huang-Wei) [inline pkg/scheduler/scheduler.go:61]: Let's add a comment here explaining the motivation, like a short duration may lead to false negative to invalidate the pod too soon, and hence cause racing conditions...

**Review by Huang-Wei — COMMENTED** [review]
LGTM. Before merging, can we brainstorm a bit more to think of any corner case that a 15-min expiration may hold the pods unnecessarily (pods should have been invalidated but haven't)?

**Comment by ahg-g** [comment]
> LGTM. Before merging, can we brainstorm a bit more to think of any corner case that a 15-min expiration may hold the pods unnecessarily (pods should have been invalidated but haven't)?

I don't have any, do you?

File: pkg/scheduler/scheduler.go (hunk 1 above)
Comment (ahg-g) [reply pkg/scheduler/scheduler.go:61]: the motivation for increasing the timeout is in the issue, not sure if it makes sense to say that it was 30s and then we increased it to 15min in the code.

Comment (ahg-g) [reply pkg/scheduler/scheduler.go:62]: I added a link to the issue.

**Comment by Huang-Wei** [comment]
> I don't have any, do you?

For 1.23, I think it's safe, in particular along with #105913.

/lgtm

But if we're going to backport, I'd suggest backporting #105913 first. Otherwise, if a "successful" bind is a false positive (due to recreated pod that holds the same name), the assumed pod would hold a stale copy, and thus may not be invalidated properly.

**Comment by ahg-g** [comment]
> > I don't have any, do you?
> 
> For 1.23, I think it's safe, in particular along with #105913.
> 
> /lgtm
> 
> But if we're going to backport, I'd suggest backporting #105913 first. Otherwise, if a "successful" bind is a false positive (due to recreated pod that holds the same name), the assumed pod would hold a stale copy, and thus may not be invalidated properly.

make sense. 

I can't think of a legitimate case, the one above is considered a bug.

**Comment by liggitt** [comment]
is this fixing a bug that existed in 1.22, or a regression in 1.23?

**Comment by ahg-g** [comment]
it existed for a long time, it got triggered recently while experimenting with really really large scale workloads.

**Comment by liggitt** [comment]
if it is not a regression, I would not recommend backporting a bug fix that requires an API change

**Comment by ahg-g** [comment]
> if it is not a regression, I would not recommend backporting a bug fix that requires an API change

but it is a bug and the api backport is backwards compatible.

**Comment by ahg-g** [comment]
to avoid backporting the api change, we can lower the timeout for the backport to strike a balance. 

The case that #105913 addresses should not be common at because aside from StatefulSet, pod names are random.

**Comment by ahg-g** [comment]
discussed offline with @liggitt, the backport will not include the api change, but will reduce the timeout to 2min

**Comment by cpanato** [comment]
@ahg-g do we need to backport this to 1.20 as well?

**Comment by ahg-g** [comment]
> @ahg-g do we need to backport this to 1.20 as well?

yup, sent https://github.com/kubernetes/kubernetes/pull/106686 as well

**Pull request merged**